tools/prepare_software_list.py

```diff
@@ -0,0 +1,290 @@
+'''
+Download / prepare / process XMPP DOAP files for the software list
+Requires: Pillow, python-slugify
+'''
+from typing import Optional
+from typing import Union
+from typing import cast
+
+from datetime import datetime
+from datetime import timedelta
+from pathlib import Path
+import json
+import os
+import shutil
+import sys
+from urllib.parse import urlparse
+from xml.etree import ElementTree as ET
+
+import requests
+from PIL import Image
+from PIL.Image import Resampling
+from slugify import slugify
+
+DOWNLOAD_PATH = Path('downloads')
+DATA_PATH = Path('data')
+STATIC_PATH = Path('static')
+LOGOS_PATH = STATIC_PATH / 'images' / 'clients'
+
+XSF_CLIENTS_LIST_URL = 'https://raw.githubusercontent.com/xsf/xmpp.org/master/' \
+    'data/clients.json'
+
+DOAP_NS = 'http://usefulinc.com/ns/doap#'
+SCHEMA_NS = 'https://schema.org/'
+RDF_RESOURCE = '{http://www.w3.org/1999/02/22-rdf-syntax-ns#}resource'
+DOAP_NAME = f'.//{{{DOAP_NS}}}name'
+DOAP_SHORTDESC = f'.//{{{DOAP_NS}}}shortdesc'
+DOAP_OS = f'.//{{{DOAP_NS}}}os'
+DOAP_LOGO = f'.//{{{SCHEMA_NS}}}logo'
+
+PLATFORMS: list[str] = [
+    'Android',
+    'iOS',
+    'Windows',
+    'macOS',
+    'Linux',
+]
+
+ENTRY_LIFETIME = timedelta(days=365)
+
+
+def initialize_directory(path: Path) -> None:
+    '''
+    Remove path (if it exists) and containing files, then recreate path
+    '''
+    if path.exists() and path.is_dir():
+        shutil.rmtree(path)
+        os.mkdir(path)
+    else:
+        os.mkdir(path)
+
+
+def download_file(url: str, path: Path) -> bool:
+    '''
+    Downloads file from url and stores it in /downloads/path
+    returns success
+    '''
+    file_request = requests.get(url)
+    if not 200 >= file_request.status_code < 400:
+        print('Error while trying to download from', url)
+        return False
+
+    with open(DOWNLOAD_PATH / path, 'wb') as file_data:
+        file_data.write(file_request.content)
```

Comment:
Please use [`iter_content`](https://docs.python-requests.org/en/latest/api/#requests.Response.iter_content) and enforce an absolute limit on the file size. We don't want a misconfigured or malicious server fill our disk or RAM (`.content` buffers the entire response in-memory).

A reasonable limit would be 10 MiB IMO.

Reply:
Good point, thanks! 10 MiB it is.